### DISCUSSION TITLE
PreInventoryMoveItemEvent

### DISCUSSION THREAD

**Discussion by NO-UUID** [PaperMC/Paper#11326] [Ideas]
### Is your feature request related to a problem?

I'm thinking it might be nice idea to add PreInventoryMoveItemEvent, which would be fired before source and target inventories are checked, so at this point server doesn't even know if it should move item, but it does know it will have to try to.

This would allow protection plugins to cancel such events before expensive itemstack checks are done, and it would also allow plugins that modify say hopper behavior to do the same and skip whole builtin logic.

Furthermore, source inventory passed in InventoryMoveItemEvent has first amount set to 1 instead of actual value, which further limits plugins.

### Describe the solution you'd like.

Add PreInventoryMoveItemEvent which is fired before any checks are done on from/to inventories.

### Describe alternatives you've considered.

Running delayed task on certain InventoryMoveItemEvent events, but I feel like that's not too great solution as processing is actually delayed by 1 tick, and there's extra performance penalty for work done before InventoryMoveItemEvent is cancelled.

**Comment by masmc05** [comment]
From my personal observation, it was never the checks being expensive, 90% of the time it is the protection plugins being extremely bad which causes lags, because there are often hundreds or thousands of hoppers transferring items each tick, and protection plugins just can't properly check all of those that fast so after 1000 checks in a tick they would still be in reasonable <1 ms

**Comment by NO-UUID** [comment]
Depends on amount of hoppers?

First, first slot is pulled, that item stack would have to be matched against every itemstack in target? Then same for 2nd slot, 3rd slot, etc... having say, 100 players, 50k hoppers across a few loaded chunks, think it would add quickly, but you're right I should check spark reports.

**Comment by electronicboy** [comment]
the source stack being sized 1 is due to the patch attempting to avoid cloning due to how expensive it is, for the root tree this is now much cheaper, the only real issue is the handling of the PDC is potentially going to cause issues here if we wanted to restore some of the cloning back now that it's supposed to be cheaper.

Adding more events inside of this tightly defined loop is fairly meh, however.

**Comment by masmc05** [comment]
@IAISI A protection plugin could potentially use <https://jd.papermc.io/paper/1.21.1/org/bukkit/block/Hopper.html#setTransferCooldown(int)> to block hoppers to protect regions, this may be the fastest and working solution

**Comment by NO-UUID** [comment]
I guess you're right on performance end of things, as well as firing too many events in such a short loop.

But, in terms of plugins/API handling hoppers it would still be an issue? And looking at the code there's not many we could change InventoryMoveItemEvent to mitigate issue of setting slot amount to 1? Not without breaking other stuff and existing plugins?

Maybe the solution would be to extend Hopper api instead? And make it so that each hopper has own transfer amount that could be modified? Not quite sure how doable that would be tho. It also might break existing plugins that just assume amount would always be 1? Then again amount wouldn't be 1 it would always be whatever is set in spigot.yml config file.

I actually went explore this trying to implement own Hopper plugin but ran into issues were amount in source inventory not being the actual amount.

**Reply by electronicboy** [reply]
The amount thing was down to aikar attempting to avoid cloning due to highly nested NBT structures, I don't think that this is a real issue anymore given mojang rewriting itemstacks, the only caveat would be the compound tag data component, but, that is generally not going to be populated to the degree that was causing issues in the past.

**Comment by NO-UUID** [comment]
It appears paper has now added HopperInventorySearchEvent, however InventoryMoveItemEvent is still "broken". Now that I thought about it some more I think, it would be way more efficient if "protection" plugins just used HopperInventorySearchEvent, set target inventory to null and set cooldown, instead of doing processing in IMIE where stacks are already being processed.

And if plugins move from IMIE to HopperInventorySearchEvent, IMIE functionality could be restored so that IMIE is fired for every slot and that could be extended even further so that item taken from source inventory is actually item passed over by getItem from IMIE. 

I've only added support for changing item amount (for even better implementation IMIE and its extended Paper's IMIE should check if not only getItem but also setItem was called and adjust items accordingly), just the change that restores full IMIE functionally and allows plugins to change amount of moved item would look like this:

![image](https://github.com/user-attachments/assets/5c93bc60-53f5-4080-b9ad-1659b6910a5d)

Why? There are plenty of plugins trying to mess with hoppers, ie change amounts, add filters whatever. With current IMIE implementation that is next to impossible, developer is left to implement own hopper ticking for special hoppers or use delayed task and do lots of processing that could be skipped with proper IMIE implementation. Both of which would obviously add way more overhead. And even if you'd be willing to eat overhead there's an issue where all this logic would break with other plugins (unless IMIE is fired again by plugin trying to implement own hopper logic, but that would create even more mess and overhead).

Another "nice" idea would probably be to add hopper transfer amount to each hopper, same way cooldowns are added, allowing plugins to change transfer amount per hopper.

Thoughts?

**Comment by snowpegeon** [comment]
I'm developing an item storage plugin.
I apologize if there are any errors in the translation as it has been passed through a translation software.

I agree with this person's opinion.
My plugin is affected by the current IMIE implementation, so I have to maintain both Spigot and Paper.
It is because Paper's IMIE implementation is "broken" and the item export process cannot be realized.

I previously asked if this was a bug and was told that it was by design.
https://github.com/PaperMC/Paper/issues/10433

As of now, Paper is significantly diverging from Spigot in terms of implementation due to the impact of the Hardfork.
In fact, a great deal of effort was required to develop the 1.21.5 plugins in order to support both Spigot and Paper.
Nevertheless, we believe that leaving this issue as it is would be a great disadvantage for future plugin developers.
Once hopper operation is implemented, I plan to switch to supporting Paper only.

Now is the perfect time to consider making hopper operation possible.

**Comment by NO-UUID** [comment]
It's not by design it's for performance reasons. But IMO behavior should be reverted to match one from Spigot, so the server admin can turn off IMIE or have it fired for all slots. 

Reasoning behind that would be, anything done with IMIE is more or less useless now since it's not fired for all items and amounts are not reliable, at this point one could ask what's the point of IMIE? Protection plugins could already use HopperInventorySearchEvent instead which is fired before inventory operations are done.

And if one were to use IMIE event to try to do inventory operations it would have to include delayed tasks which introduces more overhead, cancelling IMIE which causes issues with cooldowns not being respected and hoppers going out of Sync (since they 1 tick delayed - at least).

Personally I couldn't find any other way so I ended up patching Paper.

```
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
Date: Tue, 18 Feb 2025 22:27:13 +0100
Subject: [PATCH] Fix InventoryMoveItemEvent


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index c5e7040367d7d652e4298a4f8f8e589a2c11183f..7d00203ccf3b0f1fd2f32eccdd5a34e78014ed32 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -286,17 +286,18 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 ItemStack movedItem = origItemStack;
 
                 final int originalItemCount = origItemStack.getCount();
-                final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
+                int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
                 origItemStack.setCount(movedItemCount);
 
-                // We only need to fire the event once to give protection plugins a chance to cancel this event
-                // Because nothing uses getItem, every event call should end up the same result.
                 if (!skipPushModeEventFire) {
                     movedItem = callPushMoveEvent(destination, movedItem, hopper);
                     if (movedItem == null) { // cancelled
                         origItemStack.setCount(originalItemCount);
-                        return false;
+                        continue;
                     }
+                    movedItemCount = Math.min(movedItem.getCount(), originalItemCount);
+                    movedItem.setCount(movedItemCount);
+                    origItemStack.setCount(movedItemCount);
                 }
 
                 final ItemStack remainingItem = addItem(hopper, destination, movedItem, direction);
@@ -305,7 +306,10 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                     origItemStack = origItemStack.copy(true);
                     origItemStack.setCount(originalItemCount);
                     if (!origItemStack.isEmpty()) {
-                        origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
+                        int leftAmount = originalItemCount - movedItemCount + remainingItemCount;
+                        if(leftAmount < 0)
+                            leftAmount = 0;
+                        origItemStack.setCount(leftAmount);
                     }
                     hopper.setItem(i, origItemStack);
                     destination.setChanged();
@@ -323,7 +327,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private static boolean hopperPull(final Level level, final Hopper hopper, final Container container, ItemStack origItemStack, final int i) {
         ItemStack movedItem = origItemStack;
         final int originalItemCount = origItemStack.getCount();
-        final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
+        int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
         container.setChanged(); // original logic always marks source inv as changed even if no move happens.
         movedItem.setCount(movedItemCount);
 
@@ -331,11 +335,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             movedItem = callPullMoveEvent(hopper, container, movedItem);
             if (movedItem == null) { // cancelled
                 origItemStack.setCount(originalItemCount);
-                // Drastically improve performance by returning true.
-                // No plugin could have relied on the behavior of false as the other call
-                // site for IMIE did not exhibit the same behavior
-                return true;
+                return false;
             }
+            movedItemCount = Math.min(movedItem.getCount(), originalItemCount);
+            movedItem.setCount(movedItemCount);
+            origItemStack.setCount(movedItemCount);
         }
 
         final ItemStack remainingItem = addItem(container, hopper, movedItem, null);
@@ -344,7 +348,10 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             origItemStack = origItemStack.copy(true);
             origItemStack.setCount(originalItemCount);
             if (!origItemStack.isEmpty()) {
-                origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
+                int leftAmount = originalItemCount - movedItemCount + remainingItemCount;
+                if(leftAmount < 0)
+                    leftAmount = 0;
+                origItemStack.setCount(leftAmount);
             }
 
             ignoreBlockEntityUpdates = true;


```

```
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
Date: Sat, 29 Mar 2025 16:43:44 +0100
Subject: [PATCH] Improve Hopper performance


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 7d00203ccf3b0f1fd2f32eccdd5a34e78014ed32..3e52574130eb78819560ee72a378d9d6e1454b00 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -278,6 +278,10 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private static boolean hopperPush(final Level level, final Container destination, final Direction direction, final HopperBlockEntity hopper) {
         skipPushModeEventFire = skipHopperEvents;
         boolean foundItem = false;
+
+        org.bukkit.inventory.Inventory destinationInventory = null;
+        org.bukkit.inventory.Inventory sourceInventory = null;
+
         for (int i = 0; i < hopper.getContainerSize(); ++i) {
             final ItemStack item = hopper.getItem(i);
             if (!item.isEmpty()) {
@@ -290,7 +294,12 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 origItemStack.setCount(movedItemCount);
 
                 if (!skipPushModeEventFire) {
-                    movedItem = callPushMoveEvent(destination, movedItem, hopper);
+                    if(destinationInventory == null)
+                        destinationInventory = getInventory(destination);
+                    if(sourceInventory == null)
+                        sourceInventory = hopper.getOwner(false).getInventory();
+
+                    movedItem = callPushMoveEvent(destinationInventory, sourceInventory, movedItem, hopper);
                     if (movedItem == null) { // cancelled
                         origItemStack.setCount(originalItemCount);
                         continue;
@@ -370,10 +379,9 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     }
 
     @Nullable
-    private static ItemStack callPushMoveEvent(Container destination, ItemStack itemStack, HopperBlockEntity hopper) {
-        final org.bukkit.inventory.Inventory destinationInventory = getInventory(destination);
+    private static ItemStack callPushMoveEvent(final org.bukkit.inventory.Inventory destinationInventory, final org.bukkit.inventory.Inventory sourceInventory, ItemStack itemStack, HopperBlockEntity hopper) {
         final io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent event = new io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent(
-            hopper.getOwner(false).getInventory(),
+            sourceInventory,
             org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack),
             destinationInventory,
             true
@@ -580,6 +588,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
         for (int i : slots) {
             ItemStack item = container.getItem(i);
+            if (item.isEmpty())
+                return false;
             if (item.getCount() < item.getMaxStackSize()) {
                 return false;
             }
@@ -738,9 +748,12 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             if (item.isEmpty()) {
                 // Spigot start - SPIGOT-6693, SimpleContainer#setItem
                 ItemStack leftover = ItemStack.EMPTY; // Paper - Make hoppers respect inventory max stack size
-                if (!stack.isEmpty() && stack.getCount() > destination.getMaxStackSize()) {
-                    leftover = stack; // Paper - Make hoppers respect inventory max stack size
-                    stack = stack.split(destination.getMaxStackSize());
+                if (!stack.isEmpty()) {
+                    int destinationMaxStackSize = destination.getMaxStackSize();
+                    if(stack.getCount() > destinationMaxStackSize) {
+                        leftover = stack; // Paper - Make hoppers respect inventory max stack size
+                        stack = stack.split(destinationMaxStackSize);
+                    }
                 }
                 // Spigot end
                 ignoreBlockEntityUpdates = true; // Paper - Perf: Optimize Hoppers

```

Applied of top of Sakura 1.21.4 .jar so it might not be directly applicable but you get the idea.

**Comment by electronicboy** [comment]
The event should fire for all actual attempts of transfering an item, so long as you actually check the item; otherwise it assumes that you don't care about the stack in qyuestion and just care to prevent the operation from occuring more widely;

The amount thing can likely be reverted now, given that cloning items is cheap. I would welcome a PR to prevent the cloning and messing of the source stack size; anything else, at this moment, I'm not sure what the real issue here is. This event is stupidly deep in hot territory, and we get lambasted for the performance of this, so, any win we get here is nice.

**Reply by NO-UUID** [reply]
> The event should fire for all actual attempts of transfering an item, so long as you actually check the item; otherwise it assumes that you don't care about the stack in qyuestion and just care to prevent the operation from occuring more widely;
> 
> The amount thing can likely be reverted now, given that cloning items is cheap. I would welcome a PR to prevent the cloning and messing of the source stack size; anything else, at this moment, I'm not sure what the real issue here is. This event is stupidly deep in hot territory, and we get lambasted for the performance of this, so, any win we get here is nice.

Actually now that I think about it some more, it's not just amount issue. InventoryMoveItemEvent is stopping Item transfer after first InventoryMoveItemEvent is cancelled, it's not trying to move items until exhaustion regardless of cancelled InventoryMoveItemEvents?

Say Hopper is trying to move item into another Hopper, plugins implementing for example HopperFilters could allow some type of items to pass but not all in which case existing logic would not work so well?

**Reply by snowpegeon** [reply]
I remembered a time when a friend of mine said that InventoryMoveItemEvent in Paper was causing a problem.
<img width="754" height="887" alt="image" src="https://github.com/user-attachments/assets/baf4f83c-ac03-4e08-ae68-0335fbfedb70" />

The ItemStack obtained by event.getItem() and the ItemStack obtained by event.getSource() that the block itself has refer to the same object.
My friend suggested that if you overwrite an item obtained with getSource(), the information in getItem() will be overwritten after the event.
I can't get in touch with him right now, so I can't ask him for more details.
I hope this will be of some help in solving this problem.
